src/lib/fb.js

```diff
@@ -61,26 +79,62 @@ export function userRef() {
   return firestore.collection("users").doc(state.user.uid);
 }
 
-export function updateUrl(url) {
+/**
+ * Update's the user's row in Firestore (if signed in) with an updated URL and optional audit time.
+ *
+ * @param {string} url to update the user's row with
+ * @param {!Date} auditedOn of the most recent Lighthouse run
+ * @return {!Date} the earliest audit seen for this URL
+ */
+export async function updateUrl(url, auditedOn = null) {
   const ref = userRef();
   if (!ref) {
-    // TODO(samthor): This doesn't really inform whether the user is signed-in or not
-    return null;
+    return null; // not signed in so user has never seen this site
   }
 
-  const p = ref.set(
-    {
+  const p = firestore.runTransaction(async (transaction) => {
+    const snapshot = await transaction.get(ref);
+    const data = snapshot.data() || {};
+
+    // nb. If the userUrl matches, we can't actually just return, because Firestore demands that
+    // every document read during a transaction is written again.
+
+    const update = {
       userUrl: url,
-      userUrlUpdate: firebase.firestore.FieldValue.serverTimestamp(),
-    },
-    {merge: true},
-  );
-  p.catch((err) => {
+    };
+
+    // TODO(robdodson): Users can pass (examples for the same site):
+    //   "https://google.com/", "http://google.com", "https://www.google.com/?foo"
+    // Is it worth simply using the bare hostname here? (would still have www. vs not)
+
+    const prevSeen = (data.userUrlSeen && data.userUrlSeen[url]) || null;
```

Comment:
In firebase, what do you think of renaming the properties to:

```
users: {
  [some_user_id]: {
    currentUrl: example.com,
    previousUrls: {
      foo.com: 2019-10-31,
      ...
    }
  }
}
```

I guess as I've been reading it's been a little hard for me to quickly know what userUrl and userUrlSeen were referring to. Open to other suggestions.

Reply:
So we also use the "seen" dictionary to tell us when the current URL was first added. So "previousUrls" isn't quite accurate, but if you're happy with using it anyway, I am happy to make the change.

Reply:
what if we call it `urlHistory` ?

Reply:
or you could simplify it to just `urls` ?

```
currentUrl: foo.com,
urls: {
  foo.com: 2019-10-31,
  bar.com: 2018-05-10,
}
```

Reply:
I've renamed to `currentUrl` and `urls`.